File: cpp/ttb_turn.cpp
#include <rclcpp/rclcpp.hpp>
#include <rmw/qos_profiles.h>

#include <geometry_msgs/msg/twist.hpp>
#include <sensor_msgs/msg/imu.hpp>

#include <chrono>

using namespace std::chrono_literals;

class TurtlebotTurn : public rclcpp::Node
{
public:
  TurtlebotTurn() 
      : Node("ttb_turn")
  {
    // Create Subscribe to the IMU data topic fromm Turtlebot
    imu_sub  = this->create_subscription<sensor_msgs::msg::Imu>(
      "TTBXX/imu", 
      rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_sensor_data), rmw_qos_profile_sensor_data),
      [this](const sensor_msgs::msg::Imu &msg) { this->imu_callback(msg); }
    );

    // Create Publisher for Turtlebot velocity commands
    vel_pub = this->create_publisher<geometry_msgs::msg::Twist>("TTBXX/cmd_vel", 10);

    // Timer object that controls how often your command loop function is called
    timer = this->create_wall_timer(
      100ms,                                           // Period of rate that function is called
      [this] (void) { this->command_loop_function(); } // Which function to call
    );
  }

private:

  // Callback function for receiving Imu sensor data
  void imu_callback(const sensor_msgs::msg::Imu &msg)
  {
    angular_velocity_z = msg.angular_velocity.z;
  }

  // Function called repeatedly by node.
  void command_loop_function(void)
  {
    RCLCPP_INFO(this->get_logger(), "Angular Vel: %0.3f", angular_velocity_z);
    if(count < 50) {
      vel_cmd.angular.z = 0.4;
    } else {
      vel_cmd.angular.z = 0.0;
    }

    vel_pub->publish(vel_cmd);

    ++count;
  }

  // ---------------------------------//
  // Variable used by the Node Object //
  // ---------------------------------//

  // ROS subsriber object
  rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr imu_sub;

  // ROS publisher object
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr vel_pub;

  // Control timing control
  rclcpp::TimerBase::SharedPtr timer;

  // Velocity command message
  geometry_msgs::msg::Twist vel_cmd;

  // Variable to hold angular velocity data
  double angular_velocity_z;

  // Counter
  int count = 0;
};


int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TurtlebotTurn>());
  rclcpp::shutdown();

  return 0;
}

File: python/ttb_turn.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy

from sensor_msgs.msg import Imu
from geometry_msgs.msg import Twist

class TTBController(Node):

    def __init__(self):
        super().__init__('ttb_turn')

        # setup publisher for Twist msg to /TTB01/cmd_vel with buffer size = 10
        self.publisher_ = self.create_publisher(
            Twist,
            '/TTB01/cmd_vel',
            10
        )

        # setup subscriber to Imu msg from /TTB01/imu with buffer size = 10
        qos_profile = QoSProfile(depth=10, reliability=QoSReliabilityPolicy.BEST_EFFORT)
        self.subscriber_ = self.create_subscription(
            Imu,
            '/TTB01/imu',
            self.imu_callback,
            qos_profile
        )

        # setup controller to run at 10hz (period=.1s) and call method controller_callback
        timer_period=.1
        self.timer = self.create_timer(timer_period, self.controller_callback)

        # when count >=20, stop moving vehicle
        self.count = 0

    def controller_callback(self):
        # create msg which makes TTB speed .2 m/s and angular velocity 1.0rad/s
        # if count >=20, stop moving

        msg = Twist()
        if self.count < 20:
            msg.linear.x = .2
            msg.angular.z = 1.0
            self.count += 1

        self.publisher_.publish(msg)

    def imu_callback(self, msg):
        # print angular velocity from imu message to console
        ang_vel = msg.angular_velocity.z
        self.get_logger().info(f'Angular velocity = {ang_vel}')

def main(args=None):
    rclpy.init(args=args)

    ttb_controller = TTBController()

    rclpy.spin(ttb_controller)

    ttb_controller.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
